dsdl_compiler/README.md

```diff
@@ -0,0 +1,108 @@
+
+## Install & Integration
+To get libcanard from the git, make sure all the submodules are fetched too.
+
+`git submodules update --init --recursive`
+
+### Generating files
+`python3 libcanard_dsdlc --outdir <outdir> <dsdl-definition-uavcan-folder>`
+
+NOTE: If python2 is used, monotonic library is needed.
+
+`pip install monolith`
+
+### Using generated c-modules
+Include all or only selected message c-files to your compiler script (e.g. Makefile). Add include path to root of the <dsdl-generate-output-folder>.
+
+NOTE: compiled *.o files can't be compiled into to flat "build" directory as some files in DSDL have the same name.
+
+### Float16
+Generated structs in modules use float16 type when so specified in DSDL.
+In c-modules float16 is defined as uint16_t which can be overridden by defining FLOAT16 to something else e.g. __fp16.
+
+e.g. in Makefile
+
+`
+CFLAGS += -DFLOAT16=__fp16
+`
+
+## Using generated modules
+
+#### Encode NodeStatus-broadcast message
+```
+ #include "uavcan/protocol/NodeStatus.h"
+
+    /* Reserve memory and struct for messages */
+    uint8_t packed_uavcan_msg_buf[UAVCAN_PROTOCOL_NODESTATUS_MAX_SIZE];
+    /* MAX_SIZE comes from module header as pre-calculated */
+    uavcan_protocol_NodeStatus msg;
+
+    msg.uptime_sec = GetUptime();
+
+    msg.health = HEALTH_OK;
+    msg.mode = MODE_OPERATIONAL;
+
+    msg.sub_mode = sub_mode;
+    msg.vendor_specific_status_code = vendor_status_code;
+
+    /* Encode filled struct to packed_uavcan_msg_buf, ready to be send */
+    uint32_t len_of_packed_msg = uavcan_protocol_NodeStatusEncode(&msg, packed_uavcan_msg_buf, 0, 1);
+
+    canardBroadcast(&g_canard,
+                    UAVCAN_PROTOCOL_NODESTATUS_SIGNATURE,
+                    UAVCAN_PROTOCOL_NODESTATUS_ID,
+                    &g_bc_node_status_transfer_id,
+                    CANARD_TRANSFER_PRIORITY_MEDIUM,
+                    packed_uavcan_msg_buf,
+                    len_of_packed_msg);
+```
+
+All the *Encode calls need to have offset given as 0 and root item as 1 as parameters. This is needed, as the generated code may use offset and root item info in following *Encode calls.
+*Dynamic Array* all the dynamic arrays have also _len field, which contain the info of how many data items have been stored in to dynamic array pointer.
+
+#### Decode GetSet-request
+
+```
+    /* include header */
+    #include "uavcan/protocol/param/GetSet.h"
+
+    #define GETSETREQ_NAME_MAX_SIZE 96 // max size needed for the dynamic arrays
+    /* Reserve some memory for the dynamic arrays from the stack */
+    uint8_t buff[GETSETREQ_NAME_MAX_SIZE]; 
+    uint8_t *dyn_buf_ptr = buff;
+
+    /* Reserve struct */
+    uavcan_protocol_param_GetSetRequest get_set_req;
+
+    /* Clearing the struct is necessary! Decoding does not guarantee
+       clearing all the bytes in fields in struct. */
+    memset(&get_set_req, 0x00, sizeof(uavcan_protocol_param_GetSetRequest));
```

Comment:
Also here, from an API design, we could have the decode function clear the struct completely no? It would not cost much and make the library less "surprising to use".

Reply:
I intentionally left it out (code size), but after thinking again, I agree, it is better not to leave that choise to the user.